File: ios/SvgaModels.swift
import CoreGraphics
import Foundation
import UIKit

internal struct SvgaEntity {
    let movie: MovieEntity
    let images: [String: CGImage]
    let audioData: [String: Data]

    var byteSize: Int {
        var total = 0
        for (_, img) in images { total += img.bytesPerRow * img.height }
        for (_, data) in audioData { total += data.count }
        var totalFrames = 0
        for sprite in movie.sprites { totalFrames += sprite.frames.count }
        // FrameEntity ~80 B per frame; SpriteEntity overhead ~64 B.
        total += totalFrames * 80
        total += movie.sprites.count * 64
        return total
    }
}

internal struct MovieEntity {
    let viewBoxWidth: CGFloat
    let viewBoxHeight: CGFloat
    let fps: Int
    let frames: Int
    let sprites: [SpriteEntity]
    let audios: [AudioEntity]
}

internal struct SpriteEntity {
    let imageKey: String
    let frames: [FrameEntity]
}

internal final class FrameEntity {
    let alpha: CGFloat
    let layout: CGRect
    /// Composed transform = layout-translate × bitmap-scale × frame-transform.
    /// Immutable after parse: the parser sets this exactly once via
    /// `applyComposedTransform`, before the entity is published to the cache
    /// or the player view. Treating it as `var` previously left the door open
    /// for double-composition on cache-shared entities.
    private(set) var transform: CGAffineTransform
    let hasContent: Bool

    init(alpha: CGFloat, layout: CGRect, transform: CGAffineTransform, hasContent: Bool) {
        self.alpha = alpha
        self.layout = layout
        self.transform = transform
        self.hasContent = hasContent
    }

    /// Parser-only entry point for finalising the composed transform. Must
    /// be called at most once per FrameEntity and only before the entity is
    /// shared (i.e. before `SvgaMemoryCache.put` or `playerView.entity =`).
    func applyComposedTransform(_ t: CGAffineTransform) {
        transform = t
    }
}

internal struct AudioEntity {
    let audioKey: String
    let startFrame: Int
    let endFrame: Int
    let startTime: Int
    let totalTime: Int
}

File: android/src/main/java/com/margelo/nitro/svga/SvgaModels.kt
package com.margelo.nitro.svga

import android.graphics.Bitmap
import android.graphics.Matrix
import android.graphics.RectF
import java.util.concurrent.atomic.AtomicBoolean
import java.util.concurrent.atomic.AtomicInteger

/// Reference-counted SVGA payload.
///
/// Background: `Bitmap` holds native (off-heap) memory on every Android
/// version, but on API < 28 the native allocation is only released through
/// `Bitmap.recycle()` or finalisation. With multi-MB sprites in flight,
/// relying on finalisation OOMs low-RAM devices. We can't unconditionally
/// recycle on cache eviction either — multiple `SvgaPlayerView` instances
/// can share the same cached entity. Refcount instead: each holder
/// (`SvgaMemoryCache`, `SvgaPlayerView.entity`, `HybridSvga.entity`,
/// `SvgaSourceLoader` returns) takes a retain; recycle when the count
/// reaches zero.
///
/// Ownership convention:
///   - `SvgaMemoryCache.put`     retains; `entryRemoved` releases.
///   - `SvgaMemoryCache.get`     returns +1 (caller releases when done).
///   - `SvgaSourceLoader.loadEntity` returns +1.
///   - `SvgaPlayerView.entity` setter retains the new value, releases the prior.
///   - `HybridSvga.entity` field holds +1 ownership; assignment transfers.
///
/// All retain/release calls are thread-safe.
internal class SvgaEntity(
  val movie: MovieEntity,
  val bitmaps: Map<String, Bitmap>,
  val audioData: Map<String, ByteArray>
) {
  private val refCount = AtomicInteger(0)
  private val recycled = AtomicBoolean(false)

  val byteSize: Long by lazy { computeByteSize() }

  fun retain(): SvgaEntity {
    refCount.incrementAndGet()
    return this
  }

  fun release() {
    val newCount = refCount.decrementAndGet()
    if (newCount > 0) return
    // newCount < 0 means we've over-released — log defensively. We don't
    // throw because raising here from a finaliser-adjacent path could
    // crash the process, and the entity is already gone.
    if (newCount < 0) {
      android.util.Log.w("SvgaEntity", "release() called more times than retain()")
      return
    }
    if (recycled.compareAndSet(false, true)) {
      for (bmp in bitmaps.values) {
        // isRecycled is the cheap fast-path: skip bitmaps the host app
        // already recycled (e.g. via setImageBitmap reuse).
        if (!bmp.isRecycled) {
          try {
            bmp.recycle()
          } catch (_: Throwable) {
            // some OEM bitmaps throw on recycle — ignore; GC will reclaim
          }
        }
      }
    }
  }

  private fun computeByteSize(): Long {
    var total = 0L
    for (bmp in bitmaps.values) total += bmp.byteCount.toLong()
    for (bytes in audioData.values) total += bytes.size.toLong()
    var totalFrames = 0
    for (sprite in movie.sprites) totalFrames += sprite.frames.size
    // FrameEntity (Matrix + RectF + flags) ~80 B per frame; SpriteEntity overhead ~64 B.
    total += totalFrames.toLong() * 80L
    total += movie.sprites.size.toLong() * 64L
    return total
  }
}

internal data class MovieEntity(
  val viewBoxWidth: Float,
  val viewBoxHeight: Float,
  val fps: Int,
  val frames: Int,
  val sprites: List<SpriteEntity>,
  val audios: List<AudioEntity>
)

internal data class SpriteEntity(
  val imageKey: String,
  val frames: List<FrameEntity>
)

internal data class FrameEntity(
  val alpha: Float,
  val layout: RectF,
  val transform: Matrix,
  val hasContent: Boolean
)

internal data class AudioEntity(
  val audioKey: String,
  val startFrame: Int,
  val endFrame: Int,
  val startTime: Int,
  val totalTime: Int
)
